js/ui/grid_core/ui.grid_core.header_panel.js

```diff
@@ -48,9 +50,17 @@ const HeaderPanel = ColumnsView.inherit({
 
         this.executeAction('onToolbarPreparing', options);
 
-        if(options.toolbarOptions && !isDefined(options.toolbarOptions.visible)) {
+        if(options.toolbarOptions) {
+            if(!isDefined(options.toolbarOptions.visible)) {
+                options.toolbarOptions.visible = true;
+            }
+
             const toolbarItems = options.toolbarOptions.items;
-            options.toolbarOptions.visible = !!(toolbarItems && toolbarItems.length);
+            options.toolbarOptions.visible = !!(
+                options.toolbarOptions.visible &&
+                toolbarItems &&
+                toolbarItems.length
```

Comment:
I think that only if user does not assign visible property it should be calculated by toolbarItems.length.